### PR TITLE
Add new snippet for adding a contact with the given email address

### PR DESCRIPTION
**Pull request by ko1in1u**
Example usage:

```py
# Add a contact to a specific Google account.
ad.mbs. addGoogleContact('new.contact2@gmail.com', 'my.email@gmail.com')
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/203)

### CODE REVIEW

File: src/main/java/com/google/android/mobly/snippet/bundled/ContentProviderOperationSnippet.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import android.os.Bundle;
+import android.os.RemoteException;
+import android.provider.ContactsContract;
+import androidx.test.platform.app.InstrumentationRegistry;
+import com.google.android.mobly.snippet.Snippet;
+import com.google.android.mobly.snippet.rpc.Rpc;
+import com.google.android.mobly.snippet.rpc.RpcOptional;
+import java.util.ArrayList;
+
+/* Snippet class for operating contacts. */
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContentProviderOperationSnippet.java:33]: `ContentProviderOperation` can be used to set many things right?

If you just want to operate contact, name this `ContactSnippet`?

Like `contactAdd`, `contactRemove`?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/android/mobly/snippet/bundled/ContentProviderOperationSnippet.java (hunk 1 above)
Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContentProviderOperationSnippet.java]: Done.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    ContactsContract.CommonDataKinds.Email.CONTENT_ITEM_TYPE)
+                .withValue(ContactsContract.CommonDataKinds.Email.ADDRESS, contactEmailAddress)
+                .withValue(ContactsContract.CommonDataKinds.Email.TYPE,
+                    ContactsContract.CommonDataKinds.Email.TYPE_HOME).build());
+
+        // Apply the operations to the ContentProvider.
+        context.getContentResolver()
+            .applyBatch(ContactsContract.AUTHORITY, contentProviderOperations);
+
+        // Request a sync to the account.
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:68]: what does requesting sync mean?

let's have javadoc for how to use this

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java]: This is for syncing the new contact data to gmail account, move it to as separated function and add Javadoc.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  /**
+   * Requests an immediate synchronization of contact data for the specified Google account.
+   *
+   * <p>This method triggers a manual and expedited sync, ensuring that any changes to contacts
+   * associated with the account are promptly reflected on the device.</p>
+   *
+   * @param accountEmailAddress The email address of the Google account for which to request a
+   *     sync.
+   */
+  private void requestSync(String accountEmailAddress) {
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:83]: `requestSync` can mean many things so it's too generic a name.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            .withValue(ContactsContract.CommonDataKinds.Email.ADDRESS, contactEmailAddress)
+            .withValue(ContactsContract.CommonDataKinds.Email.TYPE,
+                ContactsContract.CommonDataKinds.Email.TYPE_HOME).build());
+
+    // Apply the operations to the ContentProvider.
+    context.getContentResolver()
+        .applyBatch(ContactsContract.AUTHORITY, contentProviderOperations);
+
+    // Request a sync to the account.
+    if (accountEmailAddress != null) {
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:69]: this step seems entirely separate from the rest of this method.
should probably just make this step a separate Rpc?

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java]: Done.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  @Rpc(description =
+      "Add a contact with the given email address. If a Google account is specified, the"
+          + " contact will be saved to that account; otherwise, it will be saved as a"
+          + " device-only contact.")
+  public void contactAdd(String contactEmailAddress, @RpcOptional String accountEmailAddress)
+      throws OperationApplicationException, RemoteException {
+    ArrayList<ContentProviderOperation> contentProviderOperations = new ArrayList<>();
+
+    // Specify where the new contact should be stored.
+    String accountType = accountEmailAddress == null ? null : GOOGLE_ACCOUNT_TYPE;
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:48]: the account type can be an email address?

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    contentProviderOperations.add(
+        ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
+            .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, accountType)
+            .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, accountEmailAddress).build());
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:52]: we allow the value here to be null?

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java (hunk 5 above)
Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:48]: Account type must be "google.com" for Google account.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java (hunk 6 above)
Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:52]: Yes, if email address is null then the contact will be saved on device.

File: src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java (hunk 5 above)
Comment (xpconanfan) [reply src/main/java/com/google/android/mobly/snippet/bundled/ContactSnippet.java:48]: Looks like this is really `addGoogleContact`?
can we just add this one specific feature without implicit behaviors of optional param?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

**Review by xpconanfan — APPROVED** [review]

**Pull request merged**